Make sure we only run the self tests once

mattcaswell opened the pull request:
Fixes #9909

levitte commented:
The fipsinstall test fails in one of the jobs...

slontis commented on providers/fips/selftest.c:
```diff
@@ -24,8 +35,55 @@
+ */
+# endif
+
+BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved);
```
why not use the same mechanism for the ini (i.e DllMain / __attribute((constructor)) ?.

Just as a reminder:
(1) Each time the fips module loads the integrity of the module needs to be checked.
(2) Once during installation the KATS need to also run after the integrity check..

(3) Self tests can be run on demand as a service after the module loads.

levitte replied:
> why not use the same mechanism for the ini (i.e DllMain / __attribute((constructor)) ?.

Because easier.  *One* codepath for the RUN_ONCE stuff, and no reason to deal with the complexities of diverse ini mechanisms.  The fini stuff is easier to deal with, since we only need to deal with platforms that don't implicitly load modules for the whole process life (implicit `RTLD_NODELETE`)...  this makes me think, though, that for those who do, we should add an atexit() call...  @mattcaswell?

slontis replied:
I would prefer to see the constructors because they are the mechanisms that are discussed in the FIPS documents.. For  the cases that dont handle this you will know because the state will not be setup when the post gets entered.. (i.e- no constructor has been called)

mattcaswell replied:
> why not use the same mechanism for the ini (i.e DllMain / __attribute((constructor)) ?.

There isn't a universally portable way of doing this on all platforms. We know how to do it for gcc/clang and Windows. But other platforms...who knows? Doing it this way it works *everywhere*.

We *could* use DllMain/__attribute((constructor)) on platforms where we know it works and fall back to RUN_ONCE on other platforms. But that just adds unnecessary complexity since RUN_ONCE universally works.

> I would prefer to see the constructors because they are the mechanisms that are discussed in the FIPS documents.

I don't think we should tie ourselves up in knots trying to do things exactly the same way as in the FIPS documents. My understanding was that that approach is not mandated.

slontis replied:
I guess we will see :)

t8m replied:
Isn't it mandated? We've got this impression from our test lab that the library constructor/destructor approach is actually mandated (if at all supported by the platform).

slontis replied:
Maybe a question for the lab. I thought this was the case also..

slontis replied:
@mattcaswell Here is a reference to the docs:
https://csrc.nist.gov/csrc/media/projects/cryptographic-module-validation-program/documents/fips140-2/fips1402ig.pdf

See Section 9.10 Power Up Tests for Software Module Libraries

levitte replied:
I jut had a read through, and sure enough, it does look mandated...
(which, btw, makes me wonder why the self test code isn't implemented as a constructor / .ini)

slontis replied:
So the design says we will just 'trigger' it from the constructor.. So it effectively just sets the fips state. We needed to defer so we can pass thru the relevant info to the self test at the right time.

levitte replied:
Ugh, too early in the morning...  I forget that we do the power on test at installation...

But then, please explain to me why we suddenly care about constructors for *this* PR, with reference to what's mandated in FIPS 140-2 IG?

slontis replied:
? We care because this PR is trying to do things with run once instead of using constructors.

levitte replied:
Why?

mattcaswell replied:
Hmmm. Reading the linked document is does appear to be mandated. But what do we do on platforms where we don't know how to do it this way? Or is FIPS just not supported on such a platform?

slontis replied:
Its not ideal - but If we know the platform we could fallback to just setting the fips_state in this case.. For now just assume we only support platforms that handle it...

mattcaswell replied:
> For now just assume we only support platforms that handle it...

Which is a little problematic - because by the time we get to attempting to compile this file, we're already building the FIPS module. We somehow need to know before we start - otherwise we'll end up with an unusable module being built

slontis replied:
Not  really seeing a issue if you can force it not to build with a config option..

levitte replied:
I still fail to understand why this i discussed *in this PR*, as if that has any impact on it.

@slontis, you wrote that what we do need to do in a DEP is to set an initial FIPS state (`FIPS_STATE_INIT`, I presume), and defer the rest to the first time `OSSL_provider_init()` is called.  Right?

As far as I can see, this PR is all about that deference, and nothing to do with setting the initial state.

What did I miss?

mattcaswell replied:
> @slontis, you wrote that what we do need to do in a DEP is to set an initial FIPS state (FIPS_STATE_INIT, I presume), and defer the rest to the first time OSSL_provider_init() is called. Right?

Is that allowed? My understanding from the IG was that it wasn't? If we're allowed to defer the running of the tests to the `OSSL_provider_init()` call, then isn't sufficient just to have the `FIPS_state` variable declared with an initialisation value. Now I'm confused.

mattcaswell replied:
> I still fail to understand why this i discussed in this PR, as if that has any impact on it.

It very much has an impact on it? The whole RUN_ONCE idea isn't needed if we always run the tests in  DEP.

levitte replied:
There's a catch: because we want to perform the default KATs *on installation*, i.e. run with the idea that *installation* is the power-on moment, we depend on the integrity test, which among others checks that the "installed" signature exists and is valid.  However, that check can't be performed earlier than on first `OSSL_provider_init()`, because that's how the installation data is passed.

This is in the design.

slontis replied:
This was discussed in the initial FIPS meeting.
The docs say that the post tests must be 'triggered' in the DEP.
Tim's interpretation of this was that 'trigger' would just mean setting the state.
This is what was decided would be done. The labs did read this design document, so we can assume it is ok.

levitte replied:
So the conclusion should then be that for *this* PR, the whole DEP thing isn't a concern, only the fini for removing the lock on platforms where this is applicable.

slontis replied:
The DEP code was also in the design doc, before it was all simplified..

slontis replied:
If the DEP code has to be there anyway, it may as well go in now and the RUN_ONCE is then not needed.

levitte replied:
... and for every platform that has a different .ini style, we need to add another piece of code, almost but not quite the same as well the others.

I don't see why it is necessary to add code complexity when the `RUN_ONCE` method works on all

mattcaswell replied:
> So the conclusion should then be that for this PR, the whole DEP thing isn't a concern,

It seems to me that the design is totally at odds to what I'm reading in the IG.

BTW, if all we need to do is set a variable value to count as the tests being "triggered" then can't that just be done in the variable declaration without a DEP at all? The OS will ensure that that variable is given that value during library load...

mattcaswell replied:
> I don't see why it is necessary to add code complexity when the RUN_ONCE method works on all

Well that was my starting point for this...but I'm being told above that the RUN_ONCE method isn't allowed by the IG. Confused.

slontis replied:
Don't assume that FIPS makes sense.
There are hoops - and you have to jump thru them.
Logic isnt going to help you here, thats all I am going to say :)

levitte replied:
> but I'm being told above that the RUN_ONCE method isn't allowed by the IG.

???????  I didn't read it like that, at all.

levitte replied:
If it turns out that we can't assume that *installation* equals *power-on*, then there's no reason whatsoever to have the KATs tests depend on the integrity test, as far as I can see.  In that case, we need to redesign the state machine that we have in the design document.

mattcaswell replied:
> ??????? I didn't read it like that, at all.

They sentences seem to be:

> These nuances are important because while a
> cryptographic module is allowed to intercept calls to a service when the power-up tests are running and
> suppress any output of results until the tests complete, a module is not allowed to initiate the tests from within
> that service. In other words, a software module implemented as a library shall not rely on calls inside any
> function exported as a supported service to initiate the power-up tests.

slontis replied:
@t-j-h care to comment?

levitte replied:
So basically, that says that you may not initiate the test via a callback through an external service (such as `RUN_ONCE`). Fair enough. But strictly speaking, this PR doesn't do that, all it does is create a lock, which doesn't initiate the test in and of itself.

mattcaswell replied:
>  But strictly speaking, this PR doesn't do that, all it does is create a lock, which doesn't initiate the test in and of itself.

No, that's true. But the rest of the function does do that. And this function is called from `OSSL_provider_init`. If that isn't an acceptable solution then it makes this whole PR kind of moot, and a new approach will be needed.

t-j-h replied:
The specific mechanism that is used does not have to follow precisely what is stated. 
Alternate implementations are permitted. You just have to work through these with the labs to make sure the underlying requirements are being met.

And remember that there are explicit limitations of what you are meant to safely trigger out of DllMain and friends according to the platform vendors. Nothing substantial is meant to happen in DllMain.

See [https://docs.microsoft.com/en-us/windows/win32/dlls/dllmain](url) and specifically:

> There are significant limits on what you can safely do in a DLL entry point. See General Best Practices for specific Windows APIs that are unsafe to call in DllMain. If you need anything but the simplest initialization then do that in an initialization function for the DLL. You can require applications to call the initialization function after DllMain has run and before they call any other functions in the DLL.


What is stated in the FIPS docs is actually precisely what is not meant to be done. The trigger can be setting a variable. And testing the variable elsewhere. That is allowed. We don't need to overly complicate things here.

mattcaswell replied:
> The trigger can be setting a variable

@t-j-h Does that have to be done as an explicit entry point? Or is it sufficient just to declare a variable with the "run tests" value, i.e. is it enough to do this:

````
static int FIPS_state = FIPS_STATE_INIT;
````
Or do we have to do this:

````
static int FIPS_state;

static __attribute__((constructor)) void init(void)
{
    FIPS_state = FIPS_STATE_INIT;
}
````

The two are identical functionally, but the latter is problematic because it is non-portable (there isn't a fully portable way of doing this AFAIK)

paulidale replied:
What about something like:

```C
static int FIPS_state = FIPS_STATE_INIT;

#ifdef HAVE_CONSTRUCTOR
static __attribute__((constructor)) void init(void)
{
    FIPS_state = FIPS_STATE_INIT;
}
#endif
```
It's a bit pointless but meets the requirement.

A better alternative might be:

```C
#ifdef HAVE_CONSTRUCTOR
static int FIPS_state;

static __attribute__((constructor)) void init(void)
{
    FIPS_state = FIPS_STATE_INIT;
}
#else
static int FIPS_state = FIPS_STATE_INIT;
#endif
```

slontis commented:
travis failure looks like a memleak

mattcaswell commented:
> travis failure looks like a memleak

It turns out this "memleak" was entirely bogus. The self test lock was being deallocated *after* mem leak checking was taking place. The solution is to temporarily disable mem leak checking while we
allocate the lock.

New commit pushed.

levitte commented:
Another reason to dump memdbg...

levitte commented on providers/fips/selftest.c:
```diff
@@ -88,10 +154,30 @@ int SELF_TEST_post(SELF_TEST_POST_PARAMS *st)
+    if (FIPS_state != FIPS_STATE_INIT) {
+        CRYPTO_THREAD_unlock(self_test_lock);
+        return 0;
+    }
```
Lines 171 to here, I don't quite understand.  Isn't it sufficient to just `CRYPTO_THREAD_write_lock(self_test_lock);`?

mattcaswell replied:
Something may have changed in between checking the state with the read lock, and re-obtaining it with a write lock. Therefore we have to check the state again.

paulidale replied:
Would wrapping these tests into their own function make sense?
Rather than duplicating the block of code.

slontis commented:
#pragma init can be used for some platforms..
See https://stackoverflow.com/questions/2459859/pragma-init-and-pragma-fini-using-gcc-compiler-on-linux.

So solaris & hpux should work..

t-j-h commented on providers/fips/selftest.c:
```diff
@@ -10,6 +10,17 @@
+#undef FIPS_MODE
+#include <internal/thread_once.h>
+#define FIPS_MODE
 #include "selftest.h"
```
I don't like this sort of approach of switching off the context to bypass a check - we should be defining an additional flag and allowing the check to pass rather than hiding our context here ...

mattcaswell replied:
I've changed this so that it sets a new define prior to the inclusion

levitte replied:
What happens if we just allow `RUN_ONCE` unconditionally in the FIPS module?

mattcaswell replied:
> What happens if we just allow RUN_ONCE unconditionally in the FIPS module?

In the vast majority of cases (all of them so far except this one) its not what you want to do. It's a bit of pandora's box wrt freeing up the memory afterwards. It turns out to be hard to do since there is no atexit() in the FIPS provider, and there is no portable way of doing a ".fini". (you may recall the long conversations we had about freeing resources (ref counting etc) when we first started implementing this stuff in the FIPS module). We only get away with it here because we take the explicit decision *not* to free up the lock if we're on a platform where we don't know how to do it.

We just avoid all of those issues if we disallow RUN_ONCE - and it turns out to work quite well without it. Instead resources are not allocated on a per-DSO basis, but on a per OPENSSL_CTX basis. We get a new OPENSSL_CTX every time OSSL_init_provider gets called. And then we free up the OPENSSL_CTX and all the associated resources when the teardown function gets called. All the code that historically was being done via RUN_ONCE can be better done as part of the OPENSSL_CTX init.

mattcaswell commented:
Fixup commit pushed addressing the feedback.

This now *requires* a DEP to be present in order for the module to successfully initialise. Compilation should succeed on all platforms regardless, but "make test" will fail on platforms where we don't know how to do the DEP. So far just done the gcc/clang detection and Windows. Not attempted the pragma thing.

Merging this PR could start causing defect reports where "make test" is failing. That might actually be a good thing in the short term because it will allow us to identify and fix those platforms where the DEP code doesn't work. In the long term (before release), if we go with this approach, we might need to suppress fips tests somehow on platforms where we can't do a DEP.

mspncp commented:
> This now requires a DEP...

What does „DEP“ stand for?

levitte commented:
Default Entry Point, i.e. .ini/.fini and similar.

levitte commented:
musl was brought up earlier, and what's being done here is probably not going to work too well there.  In https://wiki.musl-libc.org/functional-differences-from-glibc.html#Unloading-libraries, you can read for yourself that it will run the constructors (once), but not the destructors.  Shouldn't we have a fallback `atexit()` for such cases (which we already identify closely enough with `OPENSSL_USE_NODELETE`)?

mattcaswell commented:
> Shouldn't we have a fallback atexit() for such cases (which we already identify closely enough with OPENSSL_USE_NODELETE)?

Please....no....just no....[runs away screaming]....

mattcaswell commented:
For the platforms that we don't know how to clean up on, the simplest answer is just to leak the lock.

levitte commented:
> Please....no....just no....[runs away screaming]....

so I assume you'll take care of the leak issues later on then?  :wink: 

Awright, I'll leave that alone for now.

mattcaswell commented:
> so I assume you'll take care of the leak issues later on then?

This idea is to just have this one leak, and accept that it will occur on those platforms. Yes that might result in some people raising issues about leaks. I see that as preferable to opening the atexit() pandora's box! :-)

richsalz commented:
> For the platforms that we don't know how to clean up on, the simplest answer is just to leak the lock.

make a FAQ entry for it?

levitte commented on providers/fips/selftest.c:
```diff
@@ -24,8 +34,75 @@
+    CRYPTO_THREAD_lock_free(self_test_lock);
+}
+
+#endif
```
So, uhm, since you/we have now decided to go full DEP, is there any reason why we don't create the lock in the constructor?

Having that in the constructor would have you avoid two things:
1.  `RUN_ONCE`
2.  `CRYPTO_mem_ctrl()`

Also, it's a nicer symmetry...

mattcaswell replied:
Well, currently if you're on a platform where we don't know how to do then DEP everything still compiles and "works" in the sense that you get a fips.so file that will load - but it will fail in its self tests. If we move the lock creation into the constructor then such platforms would crash instead because the lock doesn't exist when they try to acquire it. The solution to that would be to try and detect very early on in the Configure stage that we can't support FIPS and automatically switch on "no-fips"...but that seems like a hard problem.

levitte commented:
All this talk about DEP and how we do all kinds of constructs to avoid it has me wonder, are we going in the wrong direction? 

We *have* knowledge of shared libraries and how to build them, it's (almost) all in `Configurations/shared-info.pl`, and there's nothing stopping us from adding necessary information for the non-gnu / non-windows cases, which would basically involve indicating an init and fini function name of our choosing where applicable. The rest would be to implement those functions where we need them and, say, have a `build.info` attribute `setinitfini` on those shared libraries and modules.

slontis commented:
> All this talk about DEP and how we do all kinds of constructs to avoid it has me wonder, are we going in the wrong direction?

This sounds a lot more complex than a few relatively simple platform specific #ifdefs.

levitte commented:
I think of it as trading one complexity for another.

slontis commented:
If its in a .pl file then only you will understand it :)

levitte commented:
Meh...

mattcaswell commented:
I agree with @slontis.

levitte commented:
...  Meh... :wink:

mattcaswell commented:
@slontis, @paulidale, @levitte: I've rebased this PR. I think it matches the approach that we have agreed. Please take another look.

levitte commented:
Are platforms we haven't written a DEP for punished for it?  I thought we came to the conclusion that such platforms would get there "through other means" and that this was permissible.  And yet, I don't see where such platforms are getting from `FIPS_STATE_INIT` to `FIPS_STATE_SELFTEST`.  What am I missing?

mattcaswell commented:
> Are platforms we haven't written a DEP for punished for it? I thought we came to the conclusion that such platforms would get there "through other means" and that this was permissible. And yet, I don't see where such platforms are getting from FIPS_STATE_INIT to FIPS_STATE_SELFTEST. What am I missing?

I think we need to avoid having it work by default just because we don't know a platform well enough to write a DEP for it. If a platform has a DEP mechanism but we don't know about it then it should not work "by magic". If there are platforms that do not have a DEP mechanism that we wish to support then we should explicitly add support for them on a platform by platform basis.

levitte commented:
Okie

mattcaswell commented:
Pushed. Thanks.
